File: Data Structures & Algorithms/permutations/submission-0.cpp
class Solution {
public:
    vector<vector<int>> perms;
    void backtrack(vector<int>& nums, vector<int>& ip)
    {
        if(ip.size() == nums.size())
        {
            perms.push_back(ip);
            return;
        }

        for(int n: nums)
        {
            if(!std::ranges::contains(ip, n))
            {
                ip.push_back(n);
                backtrack(nums, ip);
                ip.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> ip;
        backtrack(nums, ip);
        return perms;
    }
};

File: Data Structures & Algorithms/single-element-in-a-sorted-array/submission-0.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        
        while l <= r:
            mid = l + ((r - l) // 2)

            if ((mid - 1) < 0 or nums[mid - 1] != nums[mid]) and ((mid + 1) == len(nums) or nums[mid] != nums[mid + 1]):
                return nums[mid]
            
            lhs = (mid - 1) if nums[mid - 1] == nums[mid] else mid
            if lhs % 2 == 1:
                r = mid - 1
            else:
                l = mid + 1